lib/RestructureCFG/EnforceSingleExitPass.cpp

```diff
@@ -0,0 +1,171 @@
+      // `ScopeGraph` too, mind the `GT` parameter)
+      bool HasExit = false;
+      for (auto *BB : *I) {
+        using GT = llvm::GraphTraits<Scope<llvm::BasicBlock *>>;
+        auto Successors = make_range(GT::child_begin(BB), GT::child_end(BB));
+        for (auto Successor : Successors) {
+          if (!SCCNodes.contains(Successor)) {
+            HasExit = true;
+            break;
+          }
+        }
+
+        // We are not interested in SCCs that already have an exit, also on the
+        // `ScopeGraph`
+        if (HasExit)
+          break;
+      }
+
+      // If we found an SCC having an exiting edge on the `ScopeGraph`, we move
+      // on to the next SCC
+      if (HasExit)
+        continue;
+
+      // We elect the "furthest away" node, along some DFS, for the elected SCC.
+      // Due to the internal behavior of the `scc_iterator`, we can elect the
+      // first node in each `SCC` as the furthest away node on the DFS used
+      // during the `scc_iterator` itself. This would
+      BasicBlock *LastSCCNode = *I->begin();
+      Roots.push_back(LastSCCNode);
+    }
+
+    // We should have collected at least one root
+    revng_assert(Roots.size() != 0);
+
+    // 3: We add an additional `sink_block` where to redirect all the roots
+    //    found only if we have more than one root
+    if (Roots.size() <= 1) {
+      return false;
+    }
+
+    // 4: Create the new sink block
+    LLVMContext &Context = getContext(&F);
+    BasicBlock *SinkBlock = BasicBlock::Create(Context, "sink_block", &F);
+
+    // Add an `UnreachableInst` to the end of the `SinkBlock`
+    IRBuilder<> Builder(Context);
+    Builder.SetInsertPoint(SinkBlock);
+    Builder.CreateUnreachable();
+
+    // 5: We need to create a new entry block, ending with a conditional branch
+    //    (using the `true` constant as condition), whose `true` branch goes to
+    //    the original entry `BasicBlock`, and whose `false` branch goes to
+    //    newly create sink node. This is needed in order to keep the
+    //    `sink_block` alive on the CFG too, because the other edges incoming
+    //    are visible on the `ScopeGraph` only.
+    BasicBlock *OriginalEntry = &F.getEntryBlock();
+    BasicBlock *NewEntryBlock = BasicBlock::Create(Context,
+                                                   "new_entry_block",
+                                                   &F,
+                                                   OriginalEntry);
+    Builder.SetInsertPoint(NewEntryBlock);
+    Value *ConditionTrue = Builder.getTrue();
+    Builder.CreateCondBr(ConditionTrue, OriginalEntry, SinkBlock);
+
+    // 6: Connect all the exit blocks to the newly created `sink_block` with a
+    //    `scope_closer` edge
+    for (BasicBlock *Root : Roots) {
+
+      // Build a `scope_closer` edge from each identified exit block to the
+      // `sink_block`
+      ScopeGraphBuilder SGBuilder(Root->getParent());
+      SGBuilder.addScopeCloser(Root, SinkBlock);
+    }
```

Comment:
With the new logic I suggested at point 3, this should add scope closer edges only from exits that are not the elected one-true-exit.

Reply:
Yeah, I changed a bit the code here, have a look at point 7 https://github.com/revng/revng/blob/98f7b2d8804ed56f760546885bbbff6c72933843/lib/RestructureCFG/EnforceSingleExitPass.cpp#L160 and 8 https://github.com/revng/revng/blob/98f7b2d8804ed56f760546885bbbff6c72933843/lib/RestructureCFG/EnforceSingleExitPass.cpp#L172. I preferred not to mix up the logic of adding the _scope_closer_ for the _trivial_ and _non trivial_ exits, there are a couple of factorization opportunity, but I preferred super clarity here